### PR TITLE
Initial Logic and System Architecture: Create phaser scene structure with router

### PR DESCRIPTION
**Pull request by C0DECYCLE**
Repurposing this issue to add the discussed manager bases and the path sessionid transfer

### CODE REVIEW

File: src/managers/SessionManager.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import { MasterTick } from "core/masterTick";
+import { UUID } from "definitions/utils";
+import { EventEmitter } from "utilities/EventEmitter";
+import { log } from "utilities/logger";
+
+class SessionManager {
+    public readonly onSync: EventEmitter;
```
[the remark this answers is not in the document]
Comment (C0DECYCLE) [reply src/managers/SessionManager.ts:7]: The SessionManager is just a manager which has an Eventemitter which anyone can subscribe to.

File: src/managers/PlayerManager.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import { MasterTick } from "core/masterTick";
+import { EventEmitter } from "utilities/EventEmitter";
+
+class PlayerManager {
+    public readonly onSync: EventEmitter;
```
[the remark this answers is not in the document]
Comment (C0DECYCLE) [reply src/managers/PlayerManager.ts:5]: The PlayerManager is just a manager which has an Eventemitter which anyone can subscribe to.

File: src/managers/SessionManager.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+    private analyseURL(): void {
+        const sessionId: UUID = location.pathname.slice(1);
+        log(sessionId);
+    }
+
+    private beginSync(): void {
+        MasterTick.on(() => {
+            //fetch data from server and update fields
+            this.onSync.emit();
```
[the remark this answers is not in the document]
Comment (C0DECYCLE) [reply src/managers/SessionManager.ts:23]: The comment means that the server logic fetch should be implemented there and after that the the event listeners get notified that there is new data available.

File: src/managers/PlayerManager.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+    public constructor() {
+        this.onSync = new EventEmitter();
+        this.beginSync();
+    }
+
+    private beginSync(): void {
+        MasterTick.on(() => {
+            //fetch data from server and update fields
+            this.onSync.emit();
```
[the remark this answers is not in the document]
Comment (C0DECYCLE) [reply src/managers/PlayerManager.ts:15]: Event bus is something different, i will explain down below

File: src/scenes/titlescreen.ts [hunk 5]
```diff
@@ -10,6 +11,16 @@ export class TitleScreen extends Phaser.Scene {
         this.text = null;
     }
 
+    public init(): void {
+        const listener: UUID = SessionManager.onSync.on(() =>
+            log("session data synced"),
+        );
+        // on scene destroy free listener
+        this.events.on("destroy", () => {
+            SessionManager.onSync.off(listener);
```
[the remark this answers is not in the document]
Comment (C0DECYCLE) [reply src/scenes/titlescreen.ts:20]: not to confuse emitter with bus

File: src/utilities/AsyncEventEmitter.ts [hunk 6]
```diff
@@ -3,12 +3,16 @@
...
+import { AbstractEventEmitter } from "./AbstractEventEmitter";
 
 export type AsyncEventListener = (data: any) => Promise<void>;
 
 export class AsyncEventEmitter extends AbstractEventEmitter<AsyncEventListener> {
+    public constructor() {
+        super();
+    }
+
     public override once(listener: AsyncEventListener): void {
```
[the remark this answers is not in the document]
Comment (C0DECYCLE) [reply src/utilities/AsyncEventEmitter.ts:16]: You mean the hole class or the "once" method... either way there's no such logic built into there.
The once method is a variation of the on method.. the on method lets you listen to a event and the once lets you listen only once then it gets automatically removed.

**Comment by C0DECYCLE** [comment]
> It's my first time seeing event-based architecture. Please correct my comment if i understood something wrong. To sum up, the managers constantly request data from server and push the data onto one single event bus (EventEmitter). Managers are event sources. On every interval, the screens fetch the changes from the event bus because screens are subscribed to event bus.

I think you understood it differently... i wrote comments on your comments but will here explain how it works:

**Terminology:**
- Event emitter: single linear event which you can subscribe/listen to and gets emitted form time to time
- Event bus: a multi linear pattern which doesnt get used here... the only eventbus thing is the connection between react and phaser which uses a different emitter system from phaser...

**Explanation:**
There is one master event which gets emitted in a regular intervall indicating that x seconds have passed.
The managers listen for that tick and when the tick happens they will later (not implemented, only comment where to do it later) fetch data from server and because it happens periodically it acts as a sort of sync..
(Each manager has a new own event emitter which gets emitted when a new sync with the server is complete and anyone can subscribe to those emitters.)
...After a fetching is finished we then emit the sync event and anyone listening... like for example the titlescreen in this demo code is notified that new data is available and just logs a message.

Appendix:
Always remember to unsubscribe to a event if the class you subscribed from gets destroyed, otherwise massive memory leaks can occur because the automatic garbage collector wont free that class-objects memory. Thats why i, under the "on scene destroy" comment, listen when the phaser scene gets destroyed and would then unsubscribe. (because it happens that react relaunches the scene and destroys the old)

**Comment by Croxsy** [comment]
> Thanks for the explanation. So if the titlescreen needs information on players and tiles, the titlescreen will subscribe to EventEmitter in PlayerManager and TileManager.
> 
Does the Titlescreen even have to be synched? It only has to start synchronizing when clicking on create Lobby...No?

**Comment by C0DECYCLE** [comment]
> Thanks for the explanation. So if the titlescreen needs information on players and tiles, the titlescreen will subscribe to EventEmitter in PlayerManager and TileManager.

Jaein, they will subscribe to know when something changes but the data they will retrieve from a seperate getter of the manager class.

**Comment by C0DECYCLE** [comment]
> Appendix: For example, if the titlescreen gets destroyed and the client will now display lobbyscreen, the titlescreen should unsubscribe the EventEmitter in PlayerManager?

The phaser scenes dont get destroyed when switching to another scene. They will get created aka destroyed when react destroyes the entire game and recreates it... while switching scenes nothing changes.

**Comment by C0DECYCLE** [comment]
> Does the Titlescreen even have to be synched? It only has to start synchronizing when clicking on create Lobby...No?

Yess correct... this was just a example for a scene interacting with the managers. 👍

**Review by Patertuck — APPROVED** [review]

**Comment by Patertuck** [comment]
Haven't read through pr in detail, will experiment, learn and understand phaser when I use it for the main game.

**Pull request merged**